engine/rpcserver.go

```diff
@@ -1208,3 +1214,247 @@ func (s *RPCServer) GetAuditEvent(ctx context.Context, r *gctrpc.GetAuditEventRe
+		return &gctrpc.GCTScriptStatusResponse{Status: gctscript.ErrScriptingDisabled.Error()}, nil
+	}
+
+	if len(gctscript.AllVMs) < 1 {
+		return &gctrpc.GCTScriptStatusResponse{Status: "no scripts running"}, nil
+	}
+
+	resp := &gctrpc.GCTScriptStatusResponse{
+		Status: "ok",
+	}
+
+	for x := range gctscript.AllVMs {
+		resp.Scripts = append(resp.Scripts, &gctrpc.GCTScript{
+			UUID:    gctscript.AllVMs[x].ID.String(),
+			Name:    gctscript.AllVMs[x].Name,
+			NextRun: gctscript.AllVMs[x].NextRun.String(),
+		})
+	}
+
+	return resp, nil
+}
+
+func (s *RPCServer) GCTScriptQuery(ctx context.Context, r *gctrpc.GCTScriptQueryRequest) (*gctrpc.GCTScriptQueryResponse, error) {
+	if !gctscript.GCTScriptConfig.Enabled {
+		return &gctrpc.GCTScriptQueryResponse{Status: gctscript.ErrScriptingDisabled.Error()}, nil
+	}
+
+	resp := &gctrpc.GCTScriptQueryResponse{}
+
+	UUID, err := uuid.FromString(r.Script.UUID)
+	if err != nil {
+		return &gctrpc.GCTScriptQueryResponse{Status: "error"}, nil
+	}
+
+	if v, f := gctscript.AllVMs[UUID]; f {
+		resp.Status = "ok"
+		resp.Data, err = v.Read()
+		if err != nil {
+			return nil, err
+		}
+	}
+	return resp, nil
+}
+// GCTScriptExecute execute a script
+func (s *RPCServer) GCTScriptExecute(ctx context.Context, r *gctrpc.GCTScriptExecuteRequest) (*gctrpc.GCTScriptGenericResponse, error) {
+	if !gctscript.GCTScriptConfig.Enabled {
+		return &gctrpc.GCTScriptGenericResponse{Status: gctscript.ErrScriptingDisabled.Error()}, nil
+	}
+
+	if r.Script.Path == "" {
+		r.Script.Path = gctscript.ScriptPath
+	}
+
+	gctVM := gctscript.New()
+	script := filepath.Join(r.Script.Path, r.Script.Name)
+	err := gctVM.Load(script)
+	if err != nil {
+		return &gctrpc.GCTScriptGenericResponse{
+			Status: "error",
+			Data:   err.Error(),
+		}, nil
+	}
+
+	go gctVM.CompileAndRun()
+
+	return &gctrpc.GCTScriptGenericResponse{
+		Status: "ok",
+		Data:   r.Script.Name + " (" + gctVM.ID.String() + ") executed",
+	}, nil
+}
+
+// GCTScriptStop terminate a running script
+func (s *RPCServer) GCTScriptStop(ctx context.Context, r *gctrpc.GCTScriptStopRequest) (*gctrpc.GCTScriptGenericResponse, error) {
+	if !gctscript.GCTScriptConfig.Enabled {
+		return &gctrpc.GCTScriptGenericResponse{Status: gctscript.ErrScriptingDisabled.Error()}, nil
+	}
+
+	UUID, err := uuid.FromString(r.Script.UUID)
+	if err != nil {
+		return &gctrpc.GCTScriptGenericResponse{Status: "error", Data: err.Error()}, nil
+	}
+
+	if v, f := gctscript.AllVMs[UUID]; f {
+		err = v.Shutdown()
+		status := " terminated"
+		if err != nil {
+			status = err.Error()
+		}
+		return &gctrpc.GCTScriptGenericResponse{Status: "ok", Data: v.ID.String() + status}, nil
+	}
+	return &gctrpc.GCTScriptGenericResponse{Status: "error", Data: "no running script found"}, nil
+}
+
+// GCTScriptUpload upload a new script to ScriptPath
+func (s *RPCServer) GCTScriptUpload(ctx context.Context, r *gctrpc.GCTScriptUploadRequest) (*gctrpc.GCTScriptGenericResponse, error) {
```

Comment:
Just add in script validation before saving

Reply:
Validation has been added after save to verify the script is compilable

```
➜ go run ./cmd/gctcli gctscript upload "/home/andrew/dev/projects/thrashercorp/gocryptotrader-xtda/testdata/gctscript/broken.gct"            
{
 "status": "validation  failed",
 "data": "Compile Error: unresolved reference 'hello'\n\tat (main):1:1"
}
``` 

It will remove the file after failure 

There is also basic version control if a script with the same name is found it will make a copy of it into a folder version_history with current timestamp as well